### PR TITLE
Scala Stream Collector: add Azure Event Hubs sink (closes #3270)

### LINKED ISSUE #3270: Scala Stream Collector: add Azure Event Hubs sink

### PR DESCRIPTION
**Pull request by arunma**
Hi @BenFradet, Could you please have a quick look and let me know if this looks reasonable.  

The relevant collector-config looks like this : 
```
sink {
      # Choose between kinesis, googlepubsub, kafka, nsq, stdout or eventhub.
      enabled = eventhub

      # Thread pool size for EventHub API requests
      threadPoolSize = 10
      domainName = "servicebus.windows.net"

      eventHubKeys {
        goodAccessKeyName = ""
        goodAccessKey = ""
        badAccessKeyName = ""
        badAccessKey = ""
      }
```

### CODE REVIEW

**Comment by snowplowcla** [comment]
Confirmed! @arunma  has signed the Individual Contributor License Agreement. Thanks so much

File: 2-collectors/scala-stream-collector/core/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/model.scala [hunk 1]
```diff
@@ -101,6 +101,15 @@ package model {
...
   final case class Nsq(host: String, port: Int) extends SinkConfig
+  final case class EventHubKeys(goodAccessKeyName: String,
+                          goodAccessKey: String,
+                          badAccessKeyName: String,
+                          badAccessKey: String)
+  final case class EventHub(
+                             threadPoolSize: Int,
+                             domainName: String,
+                             eventHubKeys: EventHubKeys
+  ) extends SinkConfig
```
Comment (BenFradet) [inline 2-collectors/scala-stream-collector/core/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/model.scala]: :art: these two classes seems misaligned

File: 2-collectors/scala-stream-collector/eventhub/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/sinks/EventHubSink.scala [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+   * Exists so that no threads can get a reference to the EventHubSink during its construction
+   * TODO: rm scalaz \/ once 2.12
+   */
+  def createAndInitialize(
+                           eventHubConfig: EventHub,
+                           streamName: String,
+                           domainName: String,
+                           sasKeyName: String,
+                           sasKey: String,
+                           executorService: ScheduledExecutorService
```
Comment (BenFradet) [inline 2-collectors/scala-stream-collector/eventhub/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/sinks/EventHubSink.scala]: :art: this seems misaligned

File: 2-collectors/scala-stream-collector/eventhub/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/sinks/EventHubSink.scala [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+/** EventHubSink companion object with factory method */
+object EventHubSink {
+
+  /**
+   * Creates an EventHubSink
+   * Exists so that no threads can get a reference to the EventHubSink during its construction
+   * TODO: rm scalaz \/ once 2.12
+   */
+  def createAndInitialize(
+                           eventHubConfig: EventHub,
```
Comment (BenFradet) [inline 2-collectors/scala-stream-collector/eventhub/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/sinks/EventHubSink.scala]: do you need the whole config as well?

File: 2-collectors/scala-stream-collector/eventhub/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/sinks/EventHubSink.scala [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+                           sasKey: String,
+                           executorService: ScheduledExecutorService
+  ): \/[Throwable, EventHubSink] = {
+
+    val connStrBuilder = new ConnectionStringBuilder()
+      .setEndpoint( streamName, domainName)
+      .setSasKeyName(sasKeyName)
+      .setSasKey(sasKey)
+
+    val client = \/.fromTryCatch(EventHubClient.createSync(connStrBuilder.toString, executorService))
```
Comment (BenFradet) [inline 2-collectors/scala-stream-collector/eventhub/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/sinks/EventHubSink.scala]: I'm assuming there is also an async version of the client, what are the trade-offs between the two?

File: 2-collectors/scala-stream-collector/eventhub/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/sinks/EventHubSink.scala [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+  val MaxBytes = 1000000L
+
+  implicit lazy val ec = concurrent.ExecutionContext.fromExecutorService(executorService)
+
+  private val biFun = new BiFunction[Void, Throwable, Void] {
+    override def apply(v: Void, t: Throwable): Void = {
+      if (t != null) sys.error(s"Sending event to EventHub failed: ${t.getMessage}")
+      null
+    }
+  }
```
Comment (BenFradet) [inline 2-collectors/scala-stream-collector/eventhub/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/sinks/EventHubSink.scala]: In your opinion, how hard would it be to build some kind of retrying logic in this function?

File: 2-collectors/scala-stream-collector/project/Dependencies.scala [hunk 6]
```diff
@@ -51,6 +52,7 @@ object Dependencies {
     val pubsub               = "com.google.cloud"      %  "google-cloud-pubsub"    % V.pubsub
     val kafkaClients         = "org.apache.kafka"      %  "kafka-clients"          % V.kafka
     val nsqClient            = "com.snowplowanalytics" %  "nsq-java-client"        % V.nsqClient
+    val eventHub             = "com.microsoft.azure"   %   "azure-eventhubs-eph"   % V.eventHub
```
Comment (BenFradet) [inline 2-collectors/scala-stream-collector/project/Dependencies.scala]: :art: `"azure-eventhubs-eph"` is padded one character too many to the right I think

**Comment by benjben** [comment]
Closing as Ben's review remained unadressed. @arunma if you still need this feature please do not hesitate to reopen a PR in the [new repo](https://github.com/snowplow/stream-collector/) of the collector.

**Comment by arunma** [comment]
Noted, @benjben. Will do.

**Pull request closed without merging**